### PR TITLE
Add init-action script to secure the cluster (Kerberos + SSL)

### PR DESCRIPTION
**Pull request by cyxxy**

### CODE REVIEW

File: secure/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Securing Dataproc
+
+This initialization action installs the MIT distribution of the Kerberos packages, and configure Yarn, HDFS, Hive, Spark to integrate with Kerberose, which enables in-transit data encryption.
```
Comment (functicons) [inline secure/README.md]: s/Yarn/YARN, s/configure/configures

File: secure/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+It will also encrypt HTTP traffic (include Web UIs and shuffle) with SSL.
+
+## Using this initialization action
+You can use this initialization action to create a secured Dataproc cluster:
+  Using the `gcloud` command to create a new cluster with this initialization action.
+
+```bash
+gcloud dataproc clusters create <CLUSTER_NAME> \
+    --scopes cloud-platform \
```
Comment (functicons) [inline secure/README.md]: Why is this needed?

File: secure/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    --initialization-actions gs://dataproc-initialization-actions/secure/secure.sh \
+    --metadata "kms-key-uri=projects/<PROJECT_ID>/locations/global/keyRings/my-key-rings/cryptoKeys/my-key" \
```
Comment (functicons) [inline secure/README.md]: Should we parameterize key ring and key names?

File: secure/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    --metadata "keystore-uri=gs://<SECRET-BUCKET>/keystore.jks" \
+    --metadata "truststore-uri=gs://<SECRET-BUCKET>/truststore.jks" \
+    --metadata "db-password-uri=gs://<SECRET-BUCKET>/dp-password.encrypted" \
+    --metadata "root-password-uri=gs://<SECRET-BUCKET>/root-password.encrypted" \
+    --metadata "keystore-password-uri=gs://<SECRET-BUCKET>/keystore-password.encrypted"
+```
+
+All the metadata key-value pairs are required to properly secure the cluster.
+
+1. Use **keystore-uri** to specify the GCS location of the keystore file which contains the SSL certificate. It has to be in the Java KeyStore (JKS) format and when copied to VMs, renamed (if necessary) to keystore.jks. The SSL certificate should be a wildcard certificate which applies to every node in the cluster.
```
Comment (functicons) [inline secure/README.md]: Could you add some background information about keystore and truststore? Also add help for how to generate (or get) a keystore and truststore.

File: secure/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+1. Use **truststore-uri** to specify the GCS location of the truststore file. It has to be in the Java KeyStore (JKS) format and when copied to VMs, renamed (if necessary) to truststore.jks
+1. Use **keystore-password** to specify the password to the keystore and truststore files. For simplicity, the keystore password, key password, as well as the truststore password, will all be this password.
+1. Use **kms-key-uri** to specify the URI of the KMS key used to encryp various password files.
```
Comment (functicons) [inline secure/README.md]: Add link for KMS.

File: secure/README.md (hunk 1 above)
Comment (cyxxy) [reply secure/README.md]: Done.

File: secure/README.md (hunk 2 above)
Comment (cyxxy) [reply secure/README.md]: The default Dataproc scope is not enough since we need to also use KMS for password decryption. For example simplicity I just used the 'cloud-platform' scope.

File: secure/README.md (hunk 3 above)
Comment (cyxxy) [reply secure/README.md]: I believe a key-uri is enough to identify the key.

File: secure/README.md (hunk 4 above)
Comment (cyxxy) [reply secure/README.md]: I added example commands that generate keystore/truststore.

File: secure/README.md (hunk 5 above)
Comment (cyxxy) [reply secure/README.md]: Added a link, as well as example commands.

File: secure/secure.sh [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+  --key $kms_key_uri)
+readonly dataproc_bucket=$(/usr/share/google/get_metadata_value attributes/dataproc-bucket)
+readonly keystore_uri=$(/usr/share/google/get_metadata_value attributes/keystore-uri)
+readonly truststore_uri=$(/usr/share/google/get_metadata_value attributes/truststore-uri)
+readonly cluster_uuid=$(/usr/share/google/get_metadata_value attributes/dataproc-cluster-uuid)
+readonly krb5_server_mark_file="krb5-server-mark"
+
+function install_and_start_kerberos_server() {
+  DEBIAN_FRONTEND=noninteractive apt-get install -y krb5-kdc; \
+  DEBIAN_FRONTEND=noninteractive apt-get install -y krb5-admin-server || TRUE
```
Comment (karth295) [inline secure/secure.sh]: Why is it okay if this fails?

File: secure/secure.sh [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+  cat << EOF >> /etc/krb5kdc/kadm5.acl
+root *
+EOF
+
+  service krb5-admin-server restart || err 'Cannot restart Kerberos admin server'
+}
+
+function install_kerberos_client() {
+  DEBIAN_FRONTEND=noninteractive apt-get install -y krb5-user; \
+  DEBIAN_FRONTEND=noninteractive apt-get install -y krb5-config || TRUE
```
Comment (karth295) [inline secure/secure.sh]: Same question here

File: secure/secure.sh [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+
+function config_yarn_site_and_permission() {
+  # Resource Manager
+  set_property_yarn_site 'yarn.resourcemanager.keytab' "$HADOOP_CONF_DIR/yarn.keytab"
+  set_property_yarn_site 'yarn.resourcemanager.principal' "yarn/_HOST@$REALM"
+  set_property_yarn_site 'yarn.resourcemanager.hostname' "$MASTER_FQDN"
+  # Node Manager
+  set_property_yarn_site 'yarn.nodemanager.keytab' "$HADOOP_CONF_DIR/yarn.keytab"
+  set_property_yarn_site 'yarn.nodemanager.principal' "yarn/_HOST@$REALM"
+  set_property_yarn_site 'yarn.nodemanager.container-executor.class' 'org.apache.hadoop.yarn.server.nodemanager.LinuxContainerExecutor'
```
Comment (karth295) [inline secure/secure.sh]: Why use LinuxContainerExecutor (these two lines and container-executor.cfg)

File: secure/secure.sh [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+  if [[ $ROLE == 'Master' ]]; then
+    kadmin -p root -w $root_principal_password -q "addprinc -randkey spark/$FQDN"
+    kadmin -p root -w $root_principal_password -q "xst -k spark.keytab spark/$FQDN"
+    mv ./spark.keytab /etc/spark/conf/
+    chown spark:spark /etc/spark/conf/spark.keytab
+    chmod 400 /etc/spark/conf/spark.keytab
+    # Add the user 'spark' to group 'hadoop' so it can access the keystore files
+    usermod -a -G hadoop spark
+    cat << EOF >> /etc/spark/conf/spark-env.sh
+SPARK_HISTORY_OPTS="-Dspark.history.kerberos.enabled=true
```
Comment (karth295) [inline secure/secure.sh]: Just in case there are other opts already defined, `SPARK_HISTORY_OPTS="$SPARK_HISTORY_OPTS ..."`

File: secure/secure.sh [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+spark.ssl.trustStore $HADOOP_CONF_DIR/ssl/truststore.jks
+spark.ssl.keyStore $HADOOP_CONF_DIR/ssl/keystore.jks
+spark.ssl.trustStorePassword $keystore_password
+spark.ssl.keyStorePassword $keystore_password
+EOF
+  fi
+}
+
+function mark_kerberized() {
+  touch /etc/google-dataproc/kerberized
```
Comment (karth295) [inline secure/secure.sh]: Is this ever used?

File: secure/secure.sh [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+# This script installs and configures Kerberos on a Google Cloud
+# Dataproc cluster.
+
+# Do not use -x to avoid printing passwords.
```
Comment (karth295) [inline secure/secure.sh]: Since you're not using `-x`, consider adding log statements throughout

File: secure/secure.sh [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+  if [[ $ROLE == 'Master' ]]; then
+    kadmin -p root -w $root_principal_password -q "addprinc -randkey hive/$FQDN"
+    kadmin -p root -w $root_principal_password -q "xst -k hive.keytab hive/$FQDN"
+    mv ./hive.keytab /etc/hive/conf/
+    chown hive:hive /etc/hive/conf/hive.keytab
+    chmod 400 /etc/hive/conf/hive.keytab
+    # Add the user 'hive' to group 'hadoop' so it can access the keystore files
+    usermod -a -G hadoop hive
+    # HiveServer2
+    set_property_hive_site 'hive.server2.authentication' 'KERBEROS'
```
Comment (karth295) [inline secure/secure.sh]: Does `hive.server2.authentication` need to be set on workers as well so that you can run `beeline` on a worker node?

File: secure/secure.sh [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+  set_property_hdfs_site 'dfs.secondary.namenode.kerberos.principal' "hdfs/_HOST@$REALM"
+  set_property_hdfs_site 'dfs.secondary.namenode.kerberos.internal.spnego.principal' "HTTP/_HOST@$REALM"
+  # Journal Node
+  set_property_hdfs_site 'dfs.journalnode.keytab.file' "$HADOOP_CONF_DIR/hdfs.keytab"
+  set_property_hdfs_site 'dfs.journalnode.kerberos.principal' "hdfs/_HOST@$REALM"
+  set_property_hdfs_site 'dfs.journalnode.kerberos.internal.spnego.principal' "HTTP/_HOST@$REALM"
+  # Data Node
+  # 'dfs.data.transfer.protection' Enable SASL, only supported after version 2.6
+  set_property_hdfs_site 'dfs.data.transfer.protection' 'privacy'
+  unset HADOOP_SECURE_DN_USER  # Required by SASL
```
Comment (karth295) [inline secure/secure.sh]: I don't think this is necessary unless you sourced `hadoop-env.sh`.

File: secure/secure.sh [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+  config_core_site
+  config_hdfs_site
+  config_yarn_site_and_permission
+  config_mapred_site
+  copy_keystore_files
+  config_ssl_xml 'server'
+  config_ssl_xml 'client'
+  secure_hive
+  secure_spark
+  restart_services
```
Comment (karth295) [inline secure/secure.sh]: In Dataproc 1.3 I think you'll need to kerberize the app timeline server as well :/

File: secure/secure.sh (hunk 6 above)
Comment (cyxxy) [reply secure/secure.sh]: changed. If installation fails scrip will fail.

File: secure/secure.sh (hunk 7 above)
Comment (cyxxy) [reply secure/secure.sh]: ditto

File: secure/secure.sh (hunk 8 above)
Comment (cyxxy) [reply secure/secure.sh]: According to https://hadoop.apache.org/docs/r2.9.0/hadoop-yarn/hadoop-yarn-site/SecureContainer.html
"On Linux environment the secure container executor is the LinuxContainerExecutor. It uses an external program called the container-executor to launch the container. This program has the setuid access right flag set which allows it to launch the container with the permissions of the YARN application user."

File: secure/secure.sh (hunk 9 above)
Comment (cyxxy) [reply secure/secure.sh]: Done.

File: secure/secure.sh (hunk 10 above)
Comment (cyxxy) [reply secure/secure.sh]: Not now. But probably later by the Dataproc agent.

File: secure/secure.sh (hunk 11 above)
Comment (cyxxy) [reply secure/secure.sh]: Added a bunch of 'echo' commands.

File: secure/secure.sh (hunk 12 above)
Comment (cyxxy) [reply secure/secure.sh]: Done.

File: secure/secure.sh (hunk 13 above)
Comment (cyxxy) [reply secure/secure.sh]: Done.

File: secure/secure.sh (hunk 14 above)
Comment (cyxxy) [reply secure/secure.sh]: Done.

File: secure/README.md [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+    --metadata "root-password-uri=gs://<SECRET_BUCKET>/root-password.encrypted" \
+    --metadata "keystore-password-uri=gs://<SECRET_BUCKET>/keystore-password.encrypted"
+```
+
+All the metadata key-value pairs are required to properly secure the cluster.
+
+1. Use **keystore-uri** to specify the GCS location of the keystore file which contains the SSL certificate. It has to be in the Java KeyStore (JKS) format and when copied to VMs, it will be renamed (if necessary) to **keystore.jks**. The SSL certificate should be a wildcard certificate which applies to every node in the cluster.
+2. Use **truststore-uri** to specify the GCS location of the truststore file. It has to be in the Java KeyStore (JKS) format and when copied to VMs, it will be renamed (if necessary) to **truststore.jks**.
+3. Use **keystore-password** to specify the password to the keystore and truststore files. For simplicity, the keystore password, key password, as well as the truststore password, will all be this password.
+4. Use **kms-key-uri** to specify the URI of the KMS key used to encryp various password files.
```
Comment (dereklwood) [inline secure/README.md]: typo: encryp -> encrypt

File: secure/README.md [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    ```bash
+    keytool -importcert -keystore <TRUSTSTORE_FILE> \
+        -alias <ALIAS> \
+        -storepass <STORE_PASS> \
+        -file <CA_REPLY_CRT>
+    ```
+
+## Important notes
+1. This initialization action does not support Dataproc clusters in HA mode.
+2. Only a root user principal "root@`<REALM>`" will be provided (password specfied by caller). It has administrative permission to the KDC.
```
Comment (dereklwood) [inline secure/README.md]: typo: specfied -> specified

File: secure/README.md (hunk 15 above)
Comment (cyxxy) [reply secure/README.md]: fixed.

File: secure/README.md (hunk 16 above)
Comment (cyxxy) [reply secure/README.md]: fixed.

File: secure/README.md [hunk 17]
```diff
@@ -1,7 +1,8 @@
 # Securing Dataproc
 
-This initialization action installs the MIT distribution of the Kerberos packages, and configures YARN, HDFS, Hive, Spark to integrate with Kerberos, which also enables in-transit data encryption.
+This initialization action installs the MIT distribution of the Kerberos packages, and configures YARN, HDFS, Hive, Spark to integrate with Kerberose, which also enables in-transit data encryption.
```
Comment (karth295) [inline secure/README.md]: Nit: Kerberos (typo).

File: secure/secure.sh [hunk 18]
```diff
@@ -50,8 +50,11 @@ REALM=$(echo $DOMAIN | awk '{print toupper($0)}')
 MASTER="$(/usr/share/google/get_metadata_value attributes/dataproc-master)"
 MASTER_FQDN="$MASTER.$DOMAIN"
 HADOOP_CONF_DIR="/etc/hadoop/conf"
-HADOOP_VERSION=$(hadoop version | head -n 1)
 
+readonly cross_realm_trust_realm="$(/usr/share/google/get_metadata_value attributes/cross-realm-trust-realm)"
```
Comment (karth295) [inline secure/secure.sh]: I think this you need a default for these optional arguments (e.g. `echo ''`)

File: secure/secure.sh [hunk 19]
```diff
@@ -71,6 +74,12 @@ readonly keystore_password=$(gsutil cat $keystore_password_uri | \
   --ciphertext-file - \
   --plaintext-file - \
   --key $kms_key_uri)
+readonly trust_password=$(gsutil cat $cross_realm_trust_password_uri | \
```
Comment (karth295) [inline secure/secure.sh]: Similarly, this needs to be hidden behind an if statement

Comment (karth295) [inline secure/secure.sh]: Nit: here and elsewhere, our variable style is "${var_name}" -- quotes and curly braces. E.g.

```
trust_password="$(gsutil cat ${cross_realm_trust_password_uri} | \
  gcloud kms decrypt \
  --ciphertext-file - \
  --plaintext-file - \
  --key ${kms_key_uri})"
```

Comment (cyxxy) [reply secure/secure.sh]: Done.

Comment (cyxxy) [reply secure/secure.sh]: Done.

File: secure/secure.sh (hunk 18 above)
Comment (cyxxy) [reply secure/secure.sh]: Done.

File: secure/README.md (hunk 17 above)
Comment (cyxxy) [reply secure/README.md]: Done.
